### ISSUE TITLE
Verbosity level

### ISSUE THREAD

**Issue by Mdyrl7777777** [smeso/saractl#5]
Currently if global flags are set like:
`* MPROTECT,VERBOSE`
SARA logs flag for every started binary + any violations.

If flags are set like:
`* MPROTECT`
It isn't logging anything.

I would like to have option to log only violations as otherwise logs are extremely cluttered. Maybe "VERBOSE" flag should report only violations by default?

Logging something like this isn't very useful:
`SARA: WXP: '/usr/lib/sara-test/trampoline_exstack' run with flags '0x12f'.`

**Comment by smeso** [comment]
What actually happens when you set the VERBOSE flag is decided on the kernel side.
So this isn't strictly a saractl issue.
Anyway, I agree that logs at the moment are too much verbose, so I think I can change some of those printk to "DEBUG".
Thank you for your feedback.

**Comment by Mdyrl7777777** [comment]
Thanks for adjusting this!

One more thing: is it possible to rate limiting similar logging to once per app instance or once per boot? I know that's common behavior for kernel warnings.

For example currently I have this:
```
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).
```

It's much better to show `kernel: SARA: WXP: W^X in '/usr/bin/yakuake' (1295).` only once.

I can't open issue in kernel repo so I report it here.

**Comment by smeso** [comment]
Those printk are already ratelimited. Unfortunately the ratelimit facility can't take into account the originating

**Comment by smeso** [comment]
Those printk are already ratelimited.
You can customize the "ratelimit" policy via /proc/sys/kernel/printk_ratelimit and /proc/sys/kernel/printk_ratelimit_burst, be aware that this setting will apply to the whole kernel.
I wouldn't reccommend make it too much limited, because it isn't able to distinguish between printks generated from different programs.
I suggest you to silence the program that is spamming your dmesg by removing the "VERBOSE" flag from it.
A better solution would be to "fix" the program so that it doesn't violate W^X.
Given that it's a KDE program, chances are that you'll need to recompile "qtdeclarative"
without JIT support.

**Comment by Mdyrl7777777** [comment]
Ok, if it isn't possible to distinguish between printks generated from different programs then limiting it doesn't make sense.

Rebuilding all violating packages isn't feasible in binary based distro especially as even little things like `grep` print violations.

I'll build a whitelist instead. That would be second whitelist as I have already one for apps which crash.

**Comment by smeso** [comment]
> Rebuilding all violating packages isn't feasible in binary based distro especially as even little things like grep print violations.

This shouldn't happen, unless you are using "FULL" or "MMAP" flags system-wide, which is something that you shouldn't do.
MMAP or FULL should be enabled only on few specific binaries. You should stick to "MPROTECT" for anything else and it should work with almost everything.
It you still have trouble with things like grep please open another issue.

**Comment by Mdyrl7777777** [comment]
My global flags are `* MPROTECT,VERBOSE`. I don't have any troubles beside log spam for few apps but I disabled verbose flag for them as you advised. The violation for grep is W^X but it works normally:
`SARA: WXP: W^X in '/usr/bin/grep'`

**Comment by smeso** [comment]
That's strange. I don't get any warning about grep.
Can you tell me, exactly, how to reproduce the issue?
Thank you very much!

**Comment by Mdyrl7777777** [comment]
I don't know exactly, I just see it in logs. it doesn't occur for every grep usage. Maybe some other app using grep make violation and SARA thinks it's grep? Or grep is build in different way than on your system. I'm using Archlinux.

I also see violations for wget:
`SARA: WXP: W^X in '/usr/bin/wget'`

Anyway I don't see anything strange beside minor annoyance.

**Comment by smeso** [comment]
OK.
If you come up with anything more, please let me know.
Thank you again for your help.

**Comment by Mdyrl7777777** [comment]
I found it! :grinning:

`grep -P foo` cause W^X SARA violation. According to man:
```
-P, --perl-regexp
              Interpret  the  pattern  as  a Perl-compatible regular expression (PCRE).  This is experimental and grep -P may warn of unimplemented
              features.
```

**Comment by smeso** [comment]
Great! Thank you!
Apparently `libpcre` (that is used by grep when the -P option is given) can be compiled with JIT support.
On my system it's compiled without JIT, this is the reason why I don't have any problem.
If/when S.A.R.A. will be upstreamed it will be possible to fix libpcre to detect at runtime if S.A.R.A. is enabled and disable JIT compiling.
For the moment, if you don't want to recompile `libpcre` the best thing to do would be so stay without the VERBOSE flag.
Thank you again for your help.
